src/Adapter/Product/Update/ProductDuplicator.php

```diff
@@ -230,6 +271,23 @@ private function duplicateProduct(ProductId $sourceProductId, ShopConstraint $sh
         return $duplicatedProduct;
     }
 
+    /**
+     * @param ProductId $sourceProductId
+     * @param ShopConstraint $shopConstraint
+     *
+     * @return ShopId[]
+     */
+    private function getShopIdsByConstraint(ProductId $sourceProductId, ShopConstraint $shopConstraint): array
+    {
+        if ($shopConstraint->getShopId()) {
+            return [$shopConstraint->getShopId()];
+        } elseif ($shopConstraint->getShopGroupId()) {
+            return $this->productRepository->getAssociatedShopIdsFromGroup($sourceProductId, $shopConstraint->getShopGroupId());
+        }
+
+        return $this->productRepository->getAssociatedShopIds($sourceProductId);
+    }
+
```

Comment:
We already have this function in ProductMultiShoprepository. Maybe you can inject the service to use the function instead.

Reply:
Ah good call! And I already have the repository injected in this service